FROST Trusted Dealer

jesseposner opened the pull request:
This PR implements a BIP-340 compatible threshold signature system based on FROST (Flexible Round-Optimized Schnorr Threshold Signatures) with a trusted dealer. This PR does not implement distributed key generation (DKG).

### FROST Paper

This PR is based upon the FROST [paper](https://crysp.uwaterloo.ca/software/frost/) by Chelsea Komlo and Ian Goldberg and the [draft RFC](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-frost-08).

### FROST Signing BIP

This PR implements the FROST signing [BIP](https://github.com/siv2r/bip-frost-signing).

### Prior Work

This PR is patterned on the APIs, and in many instances duplicates code, from the MuSig2 implementation due to their similarities in nonce generation and signing.

real-or-random commented on include/secp256k1_frost.h:
```diff
@@ -17,6 +17,84 @@ extern "C" {
+/** Serialize a FROST share
+ *
+ *  Returns: 1 when the share could be serialized, 0 otherwise
+ *  Args:    ctx: a secp256k1 context object
```
nit: see https://github.com/bitcoin-core/secp256k1/pull/1476/commits/aa3dd5280b4a046c03bd344bfd7f1499199a1f3c for consistent wording here

real-or-random commented on include/secp256k1_frost.h:
```diff
@@ -17,6 +17,84 @@ extern "C" {
+ *  Each participant _must_ have a secure channel with the trusted dealer with
+ *  which they can transmit shares to each other.
+ *
+ *  A new seed32 _must_ be used for each key generation session. The trusted
```
nit: s/new/fresh 

By the way, we need to say similar things for musig sessions. Feel free to steal some sentences/phrasing ideas from https://github.com/bitcoin-core/secp256k1/pull/1479.

real-or-random commented on include/secp256k1_frost.h:
```diff
@@ -17,6 +17,84 @@ extern "C" {
+ *  A new seed32 _must_ be used for each key generation session. The trusted
+ *  dealer must NOT REUSE their respective seed32 again for subsequent key
+ *  generation sessions. If a trusted dealer fails to complete this session or
+ *  start a new session to generate a new key, they must NOT REUSE their
+ *  respective seed32 again, but instead generate a new one. It is recommended
```
s/again//  ("reuse again" is saying the same thing twice)

real-or-random commented on include/secp256k1_frost.h:
```diff
@@ -17,6 +17,84 @@ extern "C" {
+    const unsigned char *in32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Creates key generation shares
```
"key shares" or "key generation" shares?

real-or-random commented on include/secp256k1_frost.h:
```diff
@@ -17,6 +17,84 @@ extern "C" {
+ *  To generate a key, a trusted dealer generates a share for each other
+ *  participant.
+ *
+ *  Each participant _must_ have a secure channel with the trusted dealer with
+ *  which they can transmit shares to each other.
```
Hm, this may be slightly confusing because one could read it as "must have established a secure channel already", but that's of course not necessary. Maybe: "The trusted dealer must transmit shares over secure channels to participants."

real-or-random commented on src/modules/frost/keygen_impl.h:
```diff
@@ -0,0 +1,146 @@
+    secp256k1_sha256_write(&sha, polygen, 16);
+    secp256k1_sha256_finalize(&sha, polygen);
+
+    /* Derive share */
```
s/share/shares

real-or-random commented on src/modules/frost/keygen_impl.h:
```diff
@@ -0,0 +1,146 @@
+    for (i = 0; i < n_participants; i++) {
+        secp256k1_scalar share_i, idx;
+
+        secp256k1_scalar_clear(&share_i);
```
This function is just for nuking memory. Use `_scalar_set_int(0)` to set to zero. (We should probably have a dedicated function, but we don't have one currently.)

real-or-random commented on include/secp256k1_frost.h:
```diff
@@ -17,6 +17,84 @@ extern "C" {
+ *  Args:            ctx: pointer to a context object
+ *  Out:          shares: pointer to the key generation shares
+ *             pubshares: pointer to the public verification shares
+ *                    pk: pointer to the x-only public key
```
Should this be an x-only key? I'm not entirely sure, I would need to think about it.

jesseposner replied:
The function [conditionally inverts the shares](https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/files/a6af50e9a4dc70768108278438c391900038d980..cf31f983ff6677e8f36aec542d03f585eaeb13be#diff-65b267bfb41db84f9bf003641ec64ef27687583f9692adb7869f561f14da6d08R134) to guarantee that the private key will match the x-only key.

Since the primary use-case will be on-chain signing, I think it's a useful simplification to take care of the negation during share generation and makes it easier to reason about in [secp256k1_frost_pubkey_get](https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/commits/772a0386c1a5f250c974efa6615972cd043f7d7f#diff-65b267bfb41db84f9bf003641ec64ef27687583f9692adb7869f561f14da6d08R187) and [secp256k1_frost_pubkey_tweak](https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/commits/772a0386c1a5f250c974efa6615972cd043f7d7f#diff-65b267bfb41db84f9bf003641ec64ef27687583f9692adb7869f561f14da6d08R203).

jesseposner replied:
Upon further reflection, I'm realizing it's not a useful simplification because in practice the output of the DKG will typically be further tweaked by a BIP32 tweak and then a taproot tweak, rather than used directly.

So I now think it would be better to have the DKG output a 33-byte key rather than an x-only key.

jesseposner replied:
I've taken a stab at implementing this change, however, it requires tracking more state during signing.

Currently, `nonce_process` takes an xonly `agg_pk` and optionally a `tweak_cache` and returns a `session`. `partial_sign` takes a `session` and optionally a `tweak_cache`.

`partial_sign` only needs to handle negation logic if a `tweak_cache` is present because the DKG outputs an x-only key. If we change the DKG to output a 33-byte key, then `nonce_process` will need to take a 33-byte `agg_pk` and it will need to save a parity bit, serialized as a byte, in the `session`, so that `partial_sign` can handle the negation logic for when a `tweak_cache` is not present.

So while I don't think there's any reason for the protocol to output a 33-byte key, it seems useful to have the implementation do so, because it avoids a serialized byte in the `session` and I don't see a downside to doing so.

jesseposner replied:
I need to compare this approach to the one taken in the signing BIP. The BIP proposes that the pubshares be saved to the `session` and the 33 byte public key is derived from the pubshares.

jesseposner replied:
https://github.com/siv2r/bip-frost-signing/issues/12

jesseposner replied:
Perhaps the best reason to not output an xonly key from the DKG is that we don't want to encourage FROST DKG outputs being used directly on-chain without an unspendable script path because unlike MuSig, FROST does not randomize the group public key: https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_note-23

siv2r replied:
The signing BIP now [uses signer public shares](https://github.com/siv2r/bip-frost-signing/commit/3f23cf9f2f2177c00048782c93b90caa84b61ca8) to initialize the tweak context (name to be updated) rather than the group public key. Hence, it no longer depends on whether the group pubkey (produced by keygen) is plain or x-only.

siv2r replied:
An alternative design choice here is for `trusted_gen` to always output `tweak_cache` and optionally `group_pk`. This would be similar to how `musig_pubkey_agg` [outputs `keyagg_cache` and optionally `agg_pk`](https://github.com/BlockstreamResearch/secp256k1-zkp/blob/master/include/secp256k1_musig.h#L203-L207).

real-or-random reviewed (commented):
some initial comments, more tomorrow

real-or-random commented:
In general, I think it would be nice to get this rebase this. (Should be mostly trivial.) Also, CMake support is missing because we have this now for all other -zkp modules (https://github.com/BlockstreamResearch/secp256k1-zkp/pull/288 ), but that's currently not essential.

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -0,0 +1,650 @@
+    ARG_CHECK(n_pubnonces > 1);
+    ARG_CHECK(my_id != 0);
+    for (i = 0; i < n_pubnonces; i++) {
+        ARG_CHECK(ids[i] != 0);
```
should we also `ARG_CHECK` for `1 <= ids[i] <= max_participant`?

jesseposner replied:
The function doesn't currently take `max_participant` as an input, and I'm not sure if it's worth requiring the extra state to perform this check.

siv2r commented on include/secp256k1_frost.h:
```diff
@@ -0,0 +1,606 @@
+    const unsigned char *msg32,
+    const secp256k1_xonly_pubkey *agg_pk,
+    const unsigned char *extra_input32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(4);
```
We can also add `pubshare` as an input argument. MuSig2 hashes the individual pubkey when generating the nonces, but we don't need to make `pubshare` a mandatory argument like MuSig2 does ([more info](https://github.com/siv2r/bip-frost-signing/blob/master/design_decisions.md#no-mandatory-pk-in-nonce-generation)).

How do we decide the args supplied to the hash function for generating a nonce? Should we throw in all available signer information into the hash function? For instance, can we include the participant identifier too?

jesseposner commented on configure.ac:
```diff
@@ -470,6 +475,14 @@ if test x"$enable_module_ecdsa_adaptor" = x"yes"; then
 
+if test x"$enable_module_frost" = x"yes"; then
+  if test x"$enable_module_schnorrsig" = x"no"; then
+    AC_MSG_ERROR([Module dependency error: You have disabled the schnorrsig module explicitly, but it is required by the musig module.])
```
`%s/musig/frost`

siv2r commented on configure.ac:
```diff
@@ -579,6 +592,9 @@ else
   if test x"$enable_module_generator" = x"yes"; then
     AC_MSG_ERROR([NUMS generator module is experimental. Use --enable-experimental to allow.])
   fi
+  if test x"$enable_module_frost" = x"yes"; then
+    AC_MSG_ERROR([FROST module is experimental. Use --enable-experimental to allow.])
+  fi
```
nit: when we execute `./configure --enable-module-frost`, the following error appears:
```bash
configure: error: MuSig module is experimental. Use --enable-experimental to allow.
``` 
instead of:
```
configure: error: FROST module is experimental. Use --enable-experimental to allow.
```
----
To resolve this issue, we should reposition the FROST module check before the MuSig experimental module check (lines 486-488).

jesseposner replied:
Can you try again? I wasn't able to reproduce the issue. 

When I run `./configure --enable-module-frost` I get `configure: error: FROST module is experimental. Use --enable-experimental to allow.`.

siv2r commented on src/modules/frost/keygen_impl.h:
```diff
@@ -0,0 +1,146 @@
+    secp256k1_sha256_finalize(&sha, polygen);
+
+    /* Derive share */
+    /* See draft-irtf-cfrg-frost-08#appendix-C.1 */
```
nit: we can update this to refer to the completed RFC: https://datatracker.ietf.org/doc/rfc9591/

jesseposner replied:
https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/commits/64e8beb8d2a9b79e8a1f242da187383e785942a6

siv2r commented on src/group.h:
```diff
@@ -202,6 +202,10 @@ static void secp256k1_ge_from_bytes(secp256k1_ge *r, const unsigned char *buf);
  */
 static int secp256k1_ge_is_in_correct_subgroup(const secp256k1_ge* ge);
 
+static void secp256k1_point_save_ext(unsigned char *data, secp256k1_ge *ge);
+
+static void secp256k1_point_load_ext(secp256k1_ge *ge, const unsigned char *data);
```
The libsecp repo defines these functions [using different names](https://github.com/bitcoin-core/secp256k1/pull/1479/files#diff-28d9b92418cbddbe487fe84847993cc5da0b096363c83631a525cff25189752cR179-R191):
- `point_save_ext` -> `ge_to_bytes_ext`
- `point_load_ext` -> `ge_from_bytes_ext`

jesseposner replied:
https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/commits/9fcaf25be7ccd96f6c4da9b292daf6ff4cf10e3b

siv2r commented on include/secp256k1_frost.h:
```diff
@@ -95,6 +153,59 @@ SECP256K1_API int secp256k1_frost_shares_trusted_gen(
+ *  3. Avoid copying (or serializing) the secnonce. This reduces the possibility
+ *     that it is used more than once for signing.
+ *
+ *  Remember that nonce reuse will leak the secret key!
```
Would it be clearer to say "will leak the agg_share" instead?

jesseposner replied:
https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/commits/ae39a3a377b8b39c4ce1ad58f32115cb77e7cebb

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -0,0 +1,233 @@
+    secp256k1_sha256 sha;
+    unsigned char seed[32];
+    unsigned char i;
+    enum { n_extra_in = 4 };
```
```suggestion
    enum { N_EXTRA_IN = 4 };
```

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -0,0 +1,233 @@
+    extra_in[1] = key32;
+    extra_in[2] = pk32;
+    extra_in[3] = extra_input32;
+    for (i = 0; i < n_extra_in; i++) {
+        unsigned char len;
+        if (extra_in[i] != NULL) {
+            len = 32;
+            secp256k1_sha256_write(&sha, &len, 1);
+            secp256k1_sha256_write(&sha, extra_in[i], 32);
```
I like this approach to hashing optional arguments—it’s clearer than [musig’s approach of using a helper function](https://github.com/BlockstreamResearch/secp256k1-zkp/blob/master/src/modules/musig/session_impl.h#L303-L310). However, there's a minor issue: the signing BIP encodes the message length as 8 bytes, but here we use 1 byte.

As stated in the [BIP's noncegen:](https://github.com/siv2r/bip-frost-signing?tab=readme-ov-file#nonce-generation-1)
> - If the optional argument _m_ is not present:
>    - Let _m_prefixed = bytes(1, 0)_
>- Else:
>    - Let _m_prefixed = bytes(1, 1) || bytes(8, len(m)) || m_

jesseposner replied:
I switched over to using the MuSig approach: https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/commits/35f453dfd9690664d0cc50ca2f953bf14d584c94

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -0,0 +1,233 @@
+    return 1;
+}
+
+static void secp256k1_nonce_function_frost(secp256k1_scalar *k, const unsigned char *session_id, const unsigned char *msg32, const unsigned char *key32, const unsigned char *pk32, const unsigned char *extra_input32) {
```
Just a note for the future: when writing tests, we should skip two noncegen test vectors—[(1) empty message and (2) 38-byte message](https://github.com/siv2r/bip-frost-signing/blob/master/reference/vectors/nonce_gen_vectors.json#L15-L34)—since our nonce generation currently only supports 32-byte messages.

siv2r replied:
`secp2566k1_nonce_function_frost_helper` doesn't add the prefix of `01` when the message is present, which needs to be updated.

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -0,0 +1,233 @@
+            secp256k1_sha256_write(&sha, &len, 1);
+        }
+    }
+    secp256k1_sha256_finalize(&sha, seed);
+
+    for (i = 0; i < 2; i++) {
+        unsigned char buf[32];
+        secp256k1_sha256_initialize(&sha);
```
Our method of hashing arguments deviates from the BIP. We compute the nonces as
```math
k_i = \mathrm{H}_{\text{no-tag}}(\mathrm{H}_{\text{FROST/nonce}}(\text{optional args}) \;||\; i - 1)
```
whereas the BIP computes them as
```math
k_i = \mathrm{H}_{\text{FROST/nonce}}(\text{optional args} \;||\; i - 1)
```

jesseposner replied:
https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/commits/35f453dfd9690664d0cc50ca2f953bf14d584c94

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -230,4 +273,192 @@ int secp256k1_frost_nonce_gen(const secp256k1_context* ctx, secp256k1_frost_secn
+               the generator (by simply adding the generator to one of the
+               malicious nonces), and this does not change the winning condition
+               of the EUF-CMA game. */
+            aggnonce_pt[i] = secp256k1_ge_const_g;
```
I believe `aggnonce_pt[i]` is allowed to be infinity. The [*NonceAgg*](https://github.com/siv2r/bip-frost-signing?tab=readme-ov-file#nonce-aggregation) specification serializes them using `cybtes_ext`, so we can use `ge_to_bytes_ext` here.

Setting the nonce to the generator point only occurs with [the final nonce](https://github.com/siv2r/bip-frost-signing?tab=readme-ov-file#session-context).

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -230,4 +273,192 @@ int secp256k1_frost_nonce_gen(const secp256k1_context* ctx, secp256k1_frost_secn
+
+/* TODO: consider updating to frost-08 to address maleability at the cost of performance */
+/* See https://github.com/cfrg/draft-irtf-cfrg-frost/pull/217 */
+static int secp256k1_frost_compute_noncehash(const secp256k1_context* ctx, unsigned char *noncehash, const unsigned char *msg, const secp256k1_frost_pubnonce * const* pubnonces, size_t n_pubnonces, const unsigned char *pk32, const size_t *ids) {
```
FROST3 computes the nonce coefficient as $b = \mathrm{H}_{\text{FROST/noncecoef}}(T || \widetilde{R_1} || \widetilde{R_2} || \widetilde{X} || msg)$, where $T$ is the set of signer ids.

siv2r replied:
We can remove the `ctx` object from the function args.

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -230,4 +273,192 @@ int secp256k1_frost_nonce_gen(const secp256k1_context* ctx, secp256k1_frost_secn
+    return 1;
+}
+
+static int secp256k1_frost_nonce_process_internal(const secp256k1_context* ctx, int *fin_nonce_parity, unsigned char *fin_nonce, secp256k1_scalar *b, secp256k1_gej *aggnoncej, const unsigned char *msg, const secp256k1_frost_pubnonce * const* pubnonces, size_t n_pubnonces, const unsigned char *pk32, const size_t *ids) {
```
We can remove the `ctx` object from the function args

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -230,4 +273,192 @@ int secp256k1_frost_nonce_gen(const secp256k1_context* ctx, secp256k1_frost_secn
+    secp256k1_gej_add_ge_var(&fin_nonce_ptj, &fin_nonce_ptj, &aggnonce[0], NULL);
+    secp256k1_ge_set_gej(&fin_nonce_pt, &fin_nonce_ptj);
+
+    if (secp256k1_ge_is_infinity(&fin_nonce_pt)) {
```
The BIP sets the final nonce to the generator point if it’s infinity.

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -230,4 +273,192 @@ int secp256k1_frost_nonce_gen(const secp256k1_context* ctx, secp256k1_frost_secn
+            secp256k1_scalar_add(&session_i.s_part, &session_i.s_part, &e_tmp);
+        }
+    }
+    /* Update the challenge by multiplying the Lagrange coefficient to prepare
```
I hadn’t considered this option when writing the BIP. So, right now, it computes $\lambda_{i, T}$ during *Sign*, not in *GetSessionValue*. Happy to update the BIP if this approach is better.

siv2r commented on include/secp256k1_frost.h:
```diff
@@ -335,6 +345,135 @@ SECP256K1_API int secp256k1_frost_nonce_gen(
+    size_t n_pubnonces,
+    const unsigned char *msg32,
+    const secp256k1_xonly_pubkey *agg_pk,
+    size_t my_id,
+    const size_t *ids,
```
In the documentation, we should probably mention that [1] the `ids` array must be unique, and [2] `my_id` must be included in `ids`. Otherwise, Lagrange interpolation will return incorrect values.

We could also implement `ARG_CHECK`s for these, but that might be a lot of work.

siv2r replied:
In BIP, the _DeriveInterpolatingValue_ function fails if both these conditions aren't met.

siv2r commented on include/secp256k1_frost.h:
```diff
@@ -335,6 +345,135 @@ SECP256K1_API int secp256k1_frost_nonce_gen(
+ *  Args:         ctx: pointer to a context object
+ *  In:     pre_sig64: 64-byte pre-signature
+ *              msg32: the 32-byte message being verified
+ *             pubkey: pointer to an x-only public key to verify with
```
```suggestion
 *             agg_pk: pointer to an x-only public key to verify with
```

siv2r commented on include/secp256k1_frost.h:
```diff
@@ -335,6 +345,135 @@ SECP256K1_API int secp256k1_frost_nonce_gen(
+ *       nonce_parity: the output of `frost_nonce_parity` called with the
+ *                     session used for producing the pre-signature
+ */
+SECP256K1_API int secp256k1_frost_verify_adaptor(
```
Just pointing out that there's another possible approach here. In [single signer Schnorr adaptor](https://github.com/BlockstreamResearch/secp256k1-zkp/pull/268/files#diff-5b4fed35fac8b6fb8210bd7be8c64663ae9a227e702fb5e7d9f2a02fe2a505d3R85), we use the `extract_adaptor_point` API instead of `verify_adaptor`—[more details](https://excalidraw.com/#room=77bd1b33053edad36dff,hTvbUv8McjWYHvp-mGZv3w). Interestingly, [MuSig](https://github.com/BlockstreamResearch/secp256k1-zkp/blob/master/src/modules/musig/adaptor_impl.h) doesn’t include either `extract_adaptor_point` or `verify_adaptor` (not sure why).

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -461,4 +505,154 @@ int secp256k1_frost_nonce_process(const secp256k1_context* ctx, secp256k1_frost_
+            secp256k1_frost_partial_sign_clear(&sk, k);
+            return 0;
+        }
+        if (secp256k1_fe_is_odd(&cache_i.pk.y) != cache_i.parity_acc) {
+            secp256k1_scalar_negate(&sk, &sk);
+        }
```
We can use this comment here (taken from musig)

```
    /* Negate sk if secp256k1_fe_is_odd(&cache_i.pk.y)) XOR cache_i.parity_acc.
     * This corresponds to the line "Let d = g⋅gacc⋅d' mod n" in the
     * specification. */
```

siv2r commented on src/modules/frost/session_impl.h:
```diff
@@ -461,4 +505,154 @@ int secp256k1_frost_nonce_process(const secp256k1_context* ctx, secp256k1_frost_
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(session != NULL);
+    ARG_CHECK(partial_sigs != NULL);
+    ARG_CHECK(n_sigs > 0);
```
```suggestion
    ARG_CHECK(n_sigs > 1);
```
Should we do this instead? 
The `frost_nonce_process` currently has `ARG_CHECK(n_pubnonce > 1)`.  We might want to use the same value in both functions, either 1 or 0.

siv2r commented on examples/frost.c:
In musig's example, every function (except `main`) is `static`

siv2r commented on examples/frost.c:
```diff
@@ -0,0 +1,224 @@
+        }
+        /* Mark signer as assigned */
+        pubnonces[i] = &signer[signer_id].pubnonce;
+        /* pubkeys[i] = &signer[signer_id].pubkey; */
```
I believe this comment was meant to be removed

siv2r commented on src/modules/frost/tests_impl.h:
https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278#discussion_r1720955395 applies here too.

siv2r commented on src/modules/frost/tests_impl.h:
```diff
@@ -0,0 +1,738 @@
+         * parities */
+        frost_tweak_test();
+    }
+    for (i = 0; i < COUNT; i++) {
+        frost_multi_hop_lock_tests();
+    }
```
We can move this function call to the previous loop

siv2r commented on src/modules/frost/tests_impl.h:
```diff
@@ -0,0 +1,738 @@
+    CHECK(secp256k1_gej_is_infinity(&summed_nonces[1]));
+}
+
+int frost_memcmp_and_randomize(unsigned char *value, const unsigned char *expected, size_t len) {
```
The same function is also used in MuSig's tests, so we can make it a helper function in `src/testutil.h`.

jesseposner commented on src/modules/frost/session_impl.h:
```diff
@@ -0,0 +1,233 @@
+    enum { n_extra_in = 4 };
+    const unsigned char *extra_in[n_extra_in];
+
+    /* TODO: this doesn't have the same sidechannel resistance as the BIP340
+     * nonce function because the seckey feeds directly into SHA. */
```
I need to bring these changes over from MuSig: https://github.com/BlockstreamResearch/secp256k1-zkp/commit/206017d67d9bb8b21d5cc924ba53e1618274774c

jesseposner replied:
https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/commits/35f453dfd9690664d0cc50ca2f953bf14d584c94

jesseposner committed:
frost trusted dealer: initialize project

This commit adds the foundational configuration and building scripts
and an initial structure for the project.

jesseposner committed:
frost trusted dealer: share generation

This commit adds trusted share generation, as well as share
serialization and parsing.

jesseposner committed:
frost trusted dealer: share verification

This commit adds share verification, as well as computation of public
verification shares.

jesseposner committed:
frost trusted dealer: key tweaking

This commits add BIP-341 ("Taproot") and BIP-32 ("ordinary") public key
tweaking.

jesseposner committed:
frost trusted dealer: nonce generation

This commits adds nonce generation, as well as serialization and
parsing.

jesseposner committed:
frost trusted dealer: nonce aggregation and adaptor signatures

This commit adds nonce aggregation, as well as adaptor signatures.

jesseposner committed:
frost trusted dealer: signature generation and aggregation

This commit adds signature generation and aggregation, as well as
partial signature serialization and parsing.

jesseposner committed:
frost trusted dealer: add example file

This commit adds an example file to demonstrate how to use the module.

jesseposner committed:
frost trusted dealer: add tests

Add api tests, nonce tests, tweak tests, sha256 tag tests, and constant
time tests.

jesseposner committed:
frost trusted dealer: add documentation file

This commit adds a documentation file with detailed instructions for how
to use the module properly.

theStack committed:
frost trusted dealer: build: add CMake support

theStack reviewed (commented):
Only looked at share generation/verification so far, left a few comments below. Fwiw, I've added CMake support, feel free to pull it in: https://github.com/theStack/secp256k1-zkp/commit/3b72fbd71abafee0aa43ff0a168fd8ae7865f7a1

jesseposner committed:
Use VERIFY_CHECK for zero x-coordinate

The x-coordinate for the participant is a hash output. This x-coordinate
must not be zero, however, the probability of a sha256 output being zero
is 1 in 2^256. Thus, it is sufficient to use VERIFY_CHECK, which allows
the function to be void, which simplifies the call-sites, as suggested
by @theStack.

jesseposner committed:
Remove unused ret variable

Suggested by @theStack.

jesseposner committed:
Rename secp256k1_frost_share to secp256k1_frost_secshare

Suggested by theStack.

jesseposner committed:
Revise nonce reuse warning

Suggested by @theStack.

jesseposner committed:
Check for equality before computing indexhash

As suggested by @theStack.

jesseposner committed:
Simplify code

Co-authored-by: Sebastian Falbesoner <sebastian.falbesoner@gmail.com>

jesseposner committed:
Improve documentation

Co-authored-by: Sebastian Falbesoner <sebastian.falbesoner@gmail.com>

jesseposner committed:
wip

BEULAHEVANJALIN committed:
- examples/frost.c: zero-init secp256k1_frost_keygen_cache to avoid UB.
- tests_impl.h:
  - compute pubshares for 0..4, call frost_pubkey_gen(ids), zero pk before pubkey_get().
  - generate for 3 signers, **populate pubnonce_ptr[]**, process with ids[], and use per-signer session[].
  - **populate partial_sig_ptr[]** before aggregation; call partial_sig_agg with &session[0].
  - call nonce_parity with &session[0] (not &session).
  - generate shares/pubshares, pass ids[] into pubkey_gen; fix loop bound (i < 2) for failure branch.
  - C89: move loop variable declarations out of for() headers.

BEULAHEVANJALIN committed:
fix(frost/tests): update tests to current FROST API

- Replace string-based participant IDs with size_t indices.

- Use per-signer sessions and pass correct session to sign/verify.

- Compute pubshares with size_t index; call pubkey_gen with ids.

- Remove ids33; adjust helper signatures; clean unused vars.

- Make DKG helper self-contained; fix sign helper flow.

BEULAHEVANJALIN committed:
fix(ctime_tests, frost): migrate ctime test to index-based FROST API

The FROST API now uses size_t participant indices instead of 33-byte IDs.

Changes:
- Replace byte/string IDs with size_t indices.
- Update function calls:
  * frost_shares_gen(ctx, shares, vss_commitment, seed, t, n)
  * frost_compute_pubshare(..., /*id=*/0|1, ...)
  * frost_pubkey_gen(ctx, &cache, pubshares_ptr, 2, ids) with size_t ids[2] = {0,1}
  * frost_nonce_process(..., /*my_id=*/0, ids, ...)
- Zero-init keygen cache (`secp256k1_frost_keygen_cache cache = {0}`) to
  avoid UB from uninitialized padding.
- Drop unused variables from the old ID-bytes scheme
  (pk[2], key2, keypair2, id[2][33], id_ptr[2], size).

BEULAHEVANJALIN committed:
fix(frost tests, example): silence unused-but-set vars; correct 32-byte msg init

- tests_impl.h: keep invalid_vss_commitment/invalid_pubnonce_ptr "used" to satisfy noverify builds; add an explicit CHECK_ILLEGAL using invalid_pubnonce_ptr; drop unused sk.
- examples/frost.c: initialize 32-byte msg as a raw 32-byte array (no NUL needed), avoiding GCC's unterminated-string warning under -Werror.

Signed-off-by: BEULAHEVANJALIN <beulahebenezer777@gmail.com>

BEULAHEVANJALIN committed:
doc: fix FROST API comments and annotations

- Clarify frost_keygen_cache description to specify it holds the group public key and tweak state.
- Mark parsing functions with SECP256K1_WARN_UNUSED_RESULT.
- Remove outdated parameter from secp256k1_frost_compute_pubshare docs.
- Fix typos in comments (my_id instead of myd_id).
- Align tweak documentation with frost_pubkey_gen.
- Improve formatting of sec_adaptor32 output parameter.

No functional or ABI changes.

BEULAHEVANJALIN committed:
frost/keygen: clean up evaluate_vss, and fix naming

No functional changes.

BEULAHEVANJALIN committed:
frost: CT secnonce checks, tagged SHA midstates, batch gej->ge, tests

- const-correct secnonce save/load; CT zero via is_zero_array
- partial_sig: parse zeroes on error; serialize validates magic
- add CT ge_set_all_gej; tests vs _var and rescale checks
- nonce: batch convert, declassify per point, zero buffers
- precompute tagged SHA for aux/nonce/noncecoef + tag tests
- pubnonce parse/serialize and const cleanup

BEULAHEVANJALIN committed:
tests(frost): refactor & harden

- make helpers static; remove dead code
- loop-based tweak helper with arrays for shares/pks
- rename summed_nonces to summed_pubnonces
- add ILLEGAL-path coverage (pubkey_gen/nonce_process/nonce_gen) + share_verify OOB
- fix partial_sig size in sign helper

DarkWindman commented on src/modules/frost/session_impl.h:
```diff
@@ -0,0 +1,683 @@
+
+    secp256k1_frost_compute_noncehash_sha256_tagged(&sha);
+    /* TODO: sort by index */
+    for (i = 0; i < n_pubnonces; i++) {
```
The BIP specification defines the function `has_duplicates(id_{1..u})`. This check is crucial for the correct computation of interpolation values. However, the implementation does not enforce this verification.

DarkWindman reviewed (commented):
Hi! We are new to the secp256k1-zkp library, but we are actively researching bip-frost-signing and have noticed several potential inconsistencies. We would appreciate any feedback to our clarification.

Christewart commented on src/modules/frost/keygen_impl.h:
```diff
@@ -0,0 +1,422 @@
+    secp256k1_sha256_write(&sha, polygen, 16);
+    secp256k1_sha256_finalize(&sha, polygen);
+
+    secp256k1_frost_vss_gen(ctx, vss_commitment, polygen, threshold);
```
nit: The RFC does vss gen and share gen in inverted order.

Christewart commented on src/modules/frost/keygen_impl.h:
```diff
@@ -0,0 +1,422 @@
+    secp256k1_frost_vss_gen(ctx, vss_commitment, polygen, threshold);
+
+    for (i = 0; i < n_participants; i++) {
+        ret &= secp256k1_frost_share_gen(&shares[i], polygen, threshold, i);
```
I believe this will never modify `ret` to `0` because of https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/changes#r2795329446 , therefore validation in `secp256k1_frost_shares_gen` will always succeed.

Christewart commented on src/modules/frost/keygen_impl.h:
```diff
@@ -0,0 +1,422 @@
+        return 0;
+    }
+
+    secp256k1_scalar_negate(&share_neg, &share_i);
```
I'm following the[ RFC C.2](https://www.rfc-editor.org/rfc/rfc9591.html#appendix-C.2) and I don't understand where `L282-L286` are coming from. Can you help me understand what this is doing?

BEULAHEVANJALIN committed:
frost: fix hash_ctx API, 01 prefix bug, ret bug, cmake

BEULAHEVANJALIN committed:
cmake: add FROST experimental guard and default to OFF

FROST is not yet merged and is gated behind --enable-experimental in
configure.ac. Match that behavior in CMake by defaulting to OFF and
requiring -DSECP256K1_EXPERIMENTAL=ON, consistent with how configure.ac
treats experimental modules.

BEULAHEVANJALIN committed:
tests(frost): register tests individually with test framework

Replace the single FROST test wrapper with a tf_test_entry array using
CASE1 and REPEAT_TEST. This allows each test to be selected, executed,
and logged separately while preserving repeated test execution.

BEULAHEVANJALIN committed:
fix|refactor(frost): Fix compiler bugs and refactor function decls

- update secp256k1_ecmult_gen to secp256k1_ecmult_gen_gej func calls
- declare inputs to secp256k1_frost_*_save functions as const
- change return type of secp256k1_frost_lagrange_coefficient to void

BEULAHEVANJALIN committed:
refactor: frost/keygen_impl.h

- declare return type of `secp256k1_frost_share_gen` as void,
  as it doesn't have failure conditions
- enhance frost_share load, serialize and parse.
- modify frost_nonce_gen to return 0 if frost_share_serialize fails

BEULAHEVANJALIN committed:
refactor(frost): Clear sensistive data from stack & complete checks

BEULAHEVANJALIN committed:
refactor(frost): harden nonce/session handling

- use SHA256 tagged midstates
- explicitly clear secret material and SHA256 state
- validate pubnonce and partial signature pointers
- pass serialized secret share bytes directly to frost_nonce_gen

The nonce generation function only needs the serialized share scalar bytes,
not the secret-share magic prefix. Accepting the serialized bytes directly
avoids unnecessary secret-share deserialization and serialization error
handling in nonce generation.
